### PR TITLE
Jules - fix tests, improved diagnostics redirection etc.

### PR DESCRIPTION
**Pull request by julian-smith-artifex-com**

### CODE REVIEW

File: docs/app3.rst [hunk 1]
```diff
@@ -294,11 +294,81 @@ This design approach ensures that:
...
-Redirecting Error and Warning Messages
---------------------------------------------
-Since MuPDF version 1.16 error and warning messages can be redirected via an official plugin.
+Diagnostics
+----------------------------
 
-PyMuPDF will put error messages to `sys.stderr` prefixed with the string "mupdf:". Warnings are internally stored and can be accessed via *pymupdf.TOOLS.mupdf_warnings()*. There also is a function to empty this store.
+.. _Messages:
+
+|PyMuPDF| messages.
```
Comment (jamie-lemon) [inline docs/app3.rst:302]: pedantic - we don't usually put full stops at the end of headers

File: docs/app3.rst [hunk 2]
```diff
@@ -294,11 +294,81 @@ This design approach ensures that:
...
+two ways:
+
+*
+  Set environment variable `PYMUPDF_MESSAGE` before |PyMuPDF| is imported.
+
+*
+  Call `set_messages()`:
+
+
+MuPDF errors and warnings.
```
Comment (jamie-lemon) [inline docs/app3.rst:317]: pedantic - we don't usually put full stops at the end of headers

File: docs/recipes-common-issues-and-their-solutions.rst [hunk 3]
```diff
@@ -127,56 +127,6 @@ It features maintaining any metadata, table of contents and links contained in t
 
 
 
-How to Deal with Messages Issued by :title:`MuPDF`
```
Comment (jamie-lemon) [inline docs/recipes-common-issues-and-their-solutions.rst:130]: Shouldn't we re-write this section rather than delete it?

File: docs/tools.rst [hunk 4]
```diff
@@ -13,8 +13,9 @@ This class is a collection of utility methods and attributes, mainly around memo
 :meth:`Tools.gen_id`                   generate a unique identifier
 :meth:`Tools.store_shrink`             shrink the storables cache [#f1]_
 :meth:`Tools.mupdf_warnings`           return the accumulated MuPDF warnings
-:meth:`Tools.mupdf_display_errors`     return the accumulated MuPDF warnings
-:meth:`Tools.reset_mupdf_warnings`     empty MuPDF messages on STDOUT
+:meth:`Tools.mupdf_display_errors`     control whether MuPDF errors are displayed as messages.
+:meth:`Tools.mupdf_display_warnings`   control whether MuPDF warnings are displayed as messages.
+:meth:`Tools.reset_mupdf_warnings`     Empty MuPDF warnings/errors message buffer.
```
Comment (jamie-lemon) [inline docs/tools.rst:18]: pedantic , Empty -> empty

**Review by jamie-lemon — CHANGES_REQUESTED** [review]
Some small syntax things and a request for more in the recipes section about logging.

**Review by JorjMcKie — APPROVED** [review]

File: docs/app3.rst (hunk 1 above)
Comment (julian-smith-artifex-com) [reply docs/app3.rst:302]: Thanks, fixed in my tree.

File: docs/recipes-common-issues-and-their-solutions.rst (hunk 3 above)
Comment (julian-smith-artifex-com) [reply docs/recipes-common-issues-and-their-solutions.rst:130]: It's been rewritten and moved to section `MuPDF errors and warnings` in apps3.rst.

File: docs/tools.rst (hunk 4 above)
Comment (julian-smith-artifex-com) [reply docs/tools.rst:18]: Fixed in my tree.

**Review by jamie-lemon — APPROVED** [review]

**Pull request merged**